### PR TITLE
cresting new http server

### PR DESCRIPTION
**Pull request by boazsilverman**

### CODE REVIEW

File: src/HTTPServer/MyHttpServer.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+            // Set the response based on the result
+            String response = result ? "Word found in dictionary" : "Word not found in dictionary";
+
+            System.out.println("Sending query response: " + response);
+
+            // Send the response back to the client
+            exchange.sendResponseHeaders(200, response.length());
+            OutputStream outputStream = exchange.getResponseBody();
+            outputStream.write(response.getBytes());
```
Comment (edenfar) [inline src/HTTPServer/MyHttpServer.java:53]: return the boolean `result` rather than string

File: src/HTTPServer/MyHttpServer.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    static class ChallengeHandler implements HttpHandler {
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+            // Get the query parameter "word" from the request URL
+            String query = exchange.getRequestURI().getQuery();
+            String[] queryParams = query.split("&");
+            String word = "";
+            for (String param : queryParams) {
+                if (param.startsWith("word=")) {
+                    word = param.substring(5); // Extract the value of "word" parameter
```
Comment (edenfar) [inline src/HTTPServer/MyHttpServer.java:67]:
```suggestion
                    word = param.split("word=")[1];
```

File: src/HTTPServer/MyHttpServer.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        HttpServer server = HttpServer.create(new InetSocketAddress("localhost", 8000), 0);
+
+        // Set the context and handler for incoming requests
+        server.createContext("/query", new QueryHandler());
+        server.createContext("/challenge", new ChallengeHandler());
+
+        // Start the server
+        server.start();
+
+        System.out.println("Server is listening on port 8000");
```
Comment (edenfar) [inline src/HTTPServer/MyHttpServer.java:24]: This line is reachable? After this line the program ends

File: src/HTTPServer/MyHttpServer.java
Comment (edenfar) [inline src/HTTPServer/MyHttpServer.java:1]: All the dictionary files should move to the `HTTPServer` directory.
The hierarchy should look like:
HTTPServer
- java
  - HTTPServer
    - MyHttpServer
- resources
  - put the files here

File: src/server/java/server/Board.java [hunk 4]
```diff
@@ -140,12 +145,46 @@ public boolean boardLegal(Word w) {
     }
 
     public boolean dictionaryLegal(Word w) {
```
Comment (edenfar) [inline src/server/java/server/Board.java:147]: You should send to the server the file names as well

**Review by edenfar — CHANGES_REQUESTED** [review]

File: src/HTTPServer/MyHttpServer.java (hunk 3 above)
Comment (boazsilverman) [reply src/HTTPServer/MyHttpServer.java]: image

File: src/HTTPServer/java/MyHttpServer.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+            // Merge the word and file names into a single array
+            String[] args = new String[fileNames.size() + 1];
+            fileNames.toArray(args);
+            args[args.length - 1] = word;
+
+            // Invoke the query method of the DictionaryManager
+            boolean result = DictionaryManager.get().query(args);
+
+            // Set the response based on the result
+            String response = result ? "Word found in dictionary" : "Word not found in dictionary";
```
Comment (edenfar) [inline src/HTTPServer/java/MyHttpServer.java:64]: Return the boolean value and not string

File: src/HTTPServer/java/MyHttpServer.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    static class ChallengeHandler implements HttpHandler {
+        @Override
+        public void handle(HttpExchange exchange) throws IOException {
+            // Get the query parameter "word" from the request URL
+            String query = exchange.getRequestURI().getQuery();
+            String[] queryParams = query.split("&");
+            String word = "";
+            for (String param : queryParams) {
+                if (param.startsWith("word=")) {
+                    word = param.substring(5); // Extract the value of "word" parameter
```
Comment (edenfar) [inline src/HTTPServer/java/MyHttpServer.java:86]: The comment from the `QueryHandler` is relevant also here

File: src/server/java/server/Board.java [hunk 7]
```diff
@@ -139,13 +147,60 @@ public boolean boardLegal(Word w) {
...
+
+        try {
+
+            // Prepare the request URL for the query
+            StringJoiner fileNamesJoiner = new StringJoiner("&files=");
+            for (String fileName : queryArgs) {
+                fileNamesJoiner.add(URLEncoder.encode(fileName, StandardCharsets.UTF_8));
+            }
+
+            String requestUrl = "http://localhost:8000/query?word=" + URLEncoder.encode(wordAsString, StandardCharsets.UTF_8)
```
Comment (edenfar) [inline src/server/java/server/Board.java:163]: Are you sure it works?

File: src/server/java/server/Board.java [hunk 8]
```diff
@@ -249,7 +304,7 @@ public int tryPlaceWord(Word w) {
         if (boardLegal(test)) {
             ArrayList<Word> newWords = getWords(test);
             for (Word nw : newWords) {
-                if (dictionaryLegal(nw))
+                if (dictionaryLegal(nw,this.fileNames))
```
Comment (edenfar) [inline src/server/java/server/Board.java:307]: You should not pass `this.fileNames`, `dictionaryLegal` knows to access it

File: test/java/server/BoardTest.java [hunk 9]
```diff
@@ -22,23 +22,23 @@ public void testDictionaryLegalWordLegal() {
         Board board = new Board(new String[]{DICTIONARY_1});
         Tile[] tiles = stringToTiles("HELLO");
         Word word = new Word(tiles, 0, 0, false);
-        Assertions.assertTrue(board.dictionaryLegal(word));
+        Assertions.assertTrue(board.dictionaryLegal(word,board.getFileNames()));
```
Comment (edenfar) [inline test/java/server/BoardTest.java:25]: How the server accesses `DICTIONARY_1`?

File: src/HTTPServer/java/MyHttpServer.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
+package HTTPServer; // Update the package declaration to match your directory structure
+
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+import server.DictionaryManager;
```
Comment (edenfar) [inline src/HTTPServer/java/MyHttpServer.java:6]: Move it to the `HTTPServer/java`directory

**Review by edenfar — CHANGES_REQUESTED** [review]

File: src/server/java/server/Board.java (hunk 7 above)
Comment (boazsilverman) [reply src/server/java/server/Board.java:163]: yes

File: src/server/java/server/Board.java (hunk 8 above)
Comment (boazsilverman) [reply src/server/java/server/Board.java:307]: but there is a constructor without filemname
so they could not be there
also i wanted to keep it global

**Review by edenfar — APPROVED** [review]

**Pull request merged**